### ISSUE TITLE
bots fail to estimate distances

### ISSUE THREAD

**Issue by bmorel** [Unvanquished/Unvanquished#1672]
Bots always see absolute distance, not navmesh distance.
This notably breaks them on vertical maps like yocto, parpax is slightly less affected because the height of the roof is way bigger.

What needs to be done is to implement a function returning the lenght of the path, instead of absolute distance, and to use it almost everywhere.

**Comment by DolceTriade** [comment]
@bmorel  Can you elaborate on the following:
 - What is the implication of seeing absolute distance? Is absolute distance just the distance between two points in the unvanquished coordinate system?
 - What is navmesh distance? How can you calculate this?
 - What is the implication of navmesh distance on the ceiling?

Thanks!

**Comment by bmorel** [comment]
> Since aliens have a huge, always on, radar distance, Pythagoras's distance makes them unnecessarily aggressive, notably.
Distance have impact on the priority score of AI's targets and how much they want to retreat. In vertical maps like parpax or yocto (which are 2 official maps) this makes AIs feeding a lot.

Oh. And it also makes bots unnecessarily shy and fast to trigger their retreating code, sometimes (trying to) doing very long roads to kill instead of just trying to finish their work.
This really have a huge impact.

>> What is the implication of navmesh distance on the ceiling?

> None. 

Ok, I think I misunderstood the question here, or maybe the question is because lack of understanding, I'm not sure. Anyway, my answer is also wrong.
So:

1. navmesh on ceilings, as in, those on sky in certain maps: they are disconnected from normal, real, non-buggy, navmeshes, and hence have 0 impact AFAIK on AI's pathfinding.
2. navmeshes are not "purely 2D things". Rather, they only represent the ground bots can walk on, like a distorted 3D plane. So this feature would include the vertical parameter in the calculated distance.

**Comment by DolceTriade** [comment]
So I've been doing some research on this and I'm starting to doubt the feasibility because running this multiple times per frame will be quite expensive...
 - We need to perform a trace to go from the entities origin to ground (BotTargetToRouteTarget)
 - Do a lot of math to generate bounds (BotTargetGetBoundingBox)
 - Convert from Unv's coordinate system to the navmesh coordinate system
 - Do some graph traversal search (FindRoute)
 - Get the number of polys in the path (maybe we can just approximate distance based on the number of polys in the path, but each poly is not uniform so it is an estimate at best)
 - Traverse the path to calculate the distance

Am I missing something here @bmorel? If this is to optimize goal selection, I'm not sure its super worth it.

**Comment by bmorel** [comment]
> Do a lot of math to generate bounds (BotTargetGetBoundingBox)

Can obviously be done only when class change.

Also, IIRC recast supports partial path-finding, so if you take the problem in the other direction, that is, try to find a path that starts from the goal, to reach the bot, you can update only one part of it. I'm not sure how to do that though.
It might not be needed to do the precise calculation each frame, too, maybe once in a while wuold be enough, and the other time rely on heuristics.

>  If this is to optimize goal selection, I'm not sure its super worth it.

It's not to optimize goal selection, it's to implement support for radar and alien sense. Currently, bots with either radar or alien sense are very, very confused by it, because they see their opponent wherever it is. Combined with the fact 2 of the 3 score functions rely on distance to work properly, I have strong suspicions that this is one of the major reasons why bots really suck.
Notably, this impact how they choose their target.

Another point that is worst to mention is that currently we do a lot traces around, and I suspect this is much heavier CPU-wise than running a potentially cached A*.

**Comment by DolceTriade** [comment]
> Can obviously be done only when class change.

True, we can compute a pullthrough cache for all bounding box info for all entities.

> Also, IIRC recast supports partial path-finding,

Yes. You can do this by limiting the number of polys you pass into findRoute(). I'm not sure I understand how working backwards from the target to the bot is helpful..

>  it's to implement support for radar and alien sense

How so? Can we still use the PVS^2 algorithm we used for radar for bots (ie, if the enemy is in the view of another location that is in view by the bot is an acceptable target). This is far less expensive than all the complexities required for A* navmesh traversal.

**Comment by bmorel** [comment]
> Can we still use the PVS^2 algorithm we used for radar for bots

No idea what you're talking about, so, maybe?

The distance is really used mostly for selecting a target (it impacts the score), for selecting a heal source (same), and to keep in "memory" (you seen the `"` here?) because only the closest stuff is kept in that... huh... "memory".
I think a real distance function would help, but as always with AIs it's a lot of guts and guess. I don't know how much costly the thing would be *in practice*, and I don't know neither how much this would change the actual behaviors. I'm only persuaded that it would reduce the problems caused by vertical maps, which are caused a lot by the radar problem, but also noticeable by the rush action: they will take the closer enemy spawn as their target (and yes, this means even if their team have no reason to know it's there, but this is a different problem).

You can say the mind should be fixed before, though. It might very easily be true, but to have something sane would require more time than to implement an alternative, very likely better, distance feature, even if such a change would imply a new calibration of the various heuristics AIs use, but this task (calibration) is pretty easy (even if boring) to do: just tweak some constant magic numbers and you're done.

**Comment by necessarily-equal** [comment]
I'm afraid using navmesh for this would just be extremely costly computation-wise. You would need to keep updating a navmesh path for every enemy on every frame for every bot. Updating this for every enemy for each bot is O(n²), and could bring sudden lag spikes on a server when a bot enters a new navmesh, for example after jumping into, or out of a box.

So my traditional opinion on how to solve this was that the simple solution is to add a distance function where the z-component is weighted more, so that for example, an enemy 100m up is treated like an enemy 500m across. https://github.com/Unvanquished/Unvanquished/pull/2291 provides such a function. I think it would be an improvement anyhow over the current situation.

The PVS² algorithm sounds really neat though, if we already have an implementation we could use that. But I don't know of such an implementation.

**Comment by DolceTriade** [comment]
The code for this is in the renderer and would have to be ported to the CM system (should be a straightforward port).

visvis is precomputed here: https://github.com/DaemonEngine/Daemon/blob/aa4e8873afa6595fcb2ba1b1cebefd35853c5c93/src/engine/renderer/tr_bsp.cpp#L677

And used here: https://github.com/DaemonEngine/Daemon/blob/d2474416c67ea394160bf87f14fe141619389ea2/src/engine/renderer/tr_world.cpp#L719
